itest/rust/src/dictionary_test.rs

```diff
@@ -400,3 +425,260 @@ fn dictionary_equal() {
     assert_eq!(dict! {1: f32::NAN}, dict! {1: f32::NAN}); // yes apparently godot considers these equal
     assert_ne!(dict! {"foo": "bar"}, dict! {"bar": "foo"});
 }
+
+#[itest]
+fn dictionary_iter_create() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+
+    dictionary.iter_shared();
+    dictionary.iter_shared_as::<String, i64>();
+    dictionary.keys_shared();
+    dictionary.keys_shared_as::<String>();
+}
+
+#[itest]
+fn dictionary_iter() {
+    let dictionary = dict! {
+        "foo": 0,
+        "bar": true,
+        "baz": "foobar",
+        "nil": Variant::nil(),
+    };
+
+    let map = HashMap::<String, Variant>::from([
+        ("foo".into(), 0.to_variant()),
+        ("bar".into(), true.to_variant()),
+        ("baz".into(), "foobar".to_variant()),
+        ("nil".into(), Variant::nil()),
+    ]);
+    for (k, v) in dictionary.iter_shared() {
+        let key = String::try_from_variant(&k).unwrap();
+        assert!(map.contains_key(&key), "key = \"{key}\"");
+        assert_eq!(map[&key], v, "key = \"{key}\"");
+    }
+
+    let map2: HashMap<_, _> = dictionary
+        .iter_shared()
+        .map(|(k, v)| (String::try_from_variant(&k).unwrap(), v))
+        .collect();
+    assert_eq!(map, map2);
+    assert!(map2.contains_key("foo"));
+    assert!(map2.contains_key("bar"));
+    assert!(map2.contains_key("baz"));
+    assert!(map2.contains_key("nil"));
+}
+
+#[itest]
+fn dictionary_iter_equals_big() {
+    let dictionary: Dictionary = (0..1000).zip(0..1000).collect();
+    let map: HashMap<i64, i64> = (0..1000).zip(0..1000).collect();
+    let collected_map: HashMap<i64, i64> = dictionary
+        .iter_shared()
+        .map(|(k, v)| {
+            (
+                i64::try_from_variant(&k).unwrap(),
+                i64::try_from_variant(&v).unwrap(),
+            )
+        })
+        .collect();
+    assert_eq!(map, collected_map);
+    let collected_dictionary: Dictionary = collected_map.into_iter().collect();
+    assert_eq!(dictionary, collected_dictionary);
+}
+
+// The below tests, test to ensure no panicking happens when mutating
+// a dictionary while iterating. Though the result of doing is not
+// guaranteed.
```

Comment:
> Though the result of doing is not guaranteed.

Probably you mean that the tests passing is no guarantee that mutating while iterating is safe?

I agree that we can't be 100% certain, but I still think we can make the tests more robust. At the moment they don't contain any assertion, so which operation do you expect to panic? **If you can't find a _deliberately wrong_ implementation that would make the test fail, it's not a good test.**

Even if iteration order can be random (which I think was changed to be deterministic), can we not try to evaluate at least some properties of the result?

Reply:
I can think of many wrong implementations that would make these tests fail, mainly with accidentally dereferencing invalid variants/null pointers. but im not sure there are many good properties i could test of these would be meaningful except for "it didn't panic or do any UB while i performed the operations". most relevant properties would cause these functions to panic anyway before i get a chance to test for them.

The only thing i can think of right now would be testing that the dictionaries are valid after the iteration? which i can certainly do, but it doesn't really test the core of what i want these tests to test. though i should test that either way, because i dont think i have tested that anywhere.

As of now i can't see anything in the dictionary docs guaranteeing anything about the state of a dictionary after you mutate it mid-iteration, so i'm not sure if we should rely on an undocumented implementation detail to verify the correctness of our implementation.

Reply:
> but im not sure there are many good properties i could test of these would be meaningful except for "it didn't panic or do any UB while i performed the operations"

But panic is defined Rust behavior, and Godot methods will never do it. So what can panic? Dereferencing null pointers doesn't.

And how do you test for UB? We can run Godot compiled with address sanitizers, but at the moment we simply won't see it, except by pure chance.

What we _can_ test is that the data structures are in a valid (even if unexpected) state. You're right that Godot doesn't specify behavior, but there's probably _something_ we can assume. I'd rather have a test that breaks next year when they change implementation, than no test at all 🙂

Reply:
> But panic is defined Rust behavior, and Godot methods will never do it. So what can panic? Dereferencing null pointers doesn't.

I have some assertions in code like checking various invariants that could panic if the c++ bindings don't work the way i expect them to, and if those change behavior in the future then they would panic. 

> And how do you test for UB? We can run Godot compiled with address sanitizers, but at the moment we simply won't see it, except by pure chance.

That was the point with what i said.

> What we can test (...)

As i mentioned i will add some tests to ensure that the dictionary doesn't get messed up by the iteration with mutation. But if you don't mind relying on unspecified behavior of iteration then i can also add some tests to confirm that our iteration implementation has the same behavior as the current version of Godot, which should be true as they use the same method internally.

Reply:
> But if you don't mind relying on unspecified behavior of iteration then i can also add some tests to confirm that our iteration implementation has the same behavior as the current version of Godot, which should be true as they use the same method internally.

That sounds good to me! If we see that implementation changes too often for this to become a maintenance burden, we can reduce the testing.

And don't over-do it, having even some checks is already better than none at all 🙂 I would maybe also focus on a few checks that cover a lot of the behavior, rather than 5-10 very similar test cases. Furthermore, it's enough if the "base" versions (i.e. `keys_shared()` and `iter_shared()`) appear in these tests; all the QoL APIs on top use the same principles, and they already have their own tests.